src/dmmf/transformer.ts

```diff
@@ -87,26 +90,43 @@ function transformSchema(
  * support union types on args, but Prisma Client does.
  */
 function transformArg(arg: DMMF.SchemaArg): DmmfTypes.SchemaArg {
-  // FIXME: *Enum*Filter are currently empty
-  let inputType = arg.inputType.some((a) => a.kind === 'enum')
-    ? arg.inputType[0]
-    : arg.inputType.find((a) => a.kind === 'object')!
-
-  if (!inputType) {
-    inputType = arg.inputType[0]
-  }
+  const inputType = flattenUnionOfSchemaArg(arg.inputTypes)
 
   return {
     name: arg.name,
     inputType: {
-      ...inputType,
       type: getReturnTypeName(inputType.type),
+      isList: inputType.isList,
+      kind: inputType.kind,
+      isNullable: arg.isNullable,
+      isRequired: arg.isRequired,
     },
     // FIXME Why?
     isRelationFilter: undefined,
   }
 }
 
+/**
+ * Trying getting the "broadest" member of the union type
+ * Intentionally ignoring the `<Model>RelationFilter` for now and using the `<Model>WhereInput` instead to avoid making a breaking change
+ */
+function flattenUnionOfSchemaArg(inputTypes: DMMF.SchemaArgInputType[]): DMMF.SchemaArgInputType {
+  return (
+    // [<Model>WhereInput]
+    inputTypes.find(
+      (a) => a.kind === 'object' && a.isList == true && getReturnTypeName(a.type).endsWith('WhereInput')
+    ) ??
+    // <Model>WhereInput
+    inputTypes.find((a) => a.kind === 'object' && getReturnTypeName(a.type).endsWith('WhereInput')) ??
+    // [<AnyType>]
+    inputTypes.find((a) => a.kind === 'object' && a.isList === true) ??
+    // AnyType
+    inputTypes.find((a) => a.kind === 'object') ??
+    // fallback to the first member of the union
+    inputTypes[0]
+  )
+}
```

Comment:
Things are starting to get a bit hairy. I've come up with some heuristic to prevent some more breaking changes. We're limited to one right now

Reply:
A lot of this code to me is magical, maybe code comments are worthwhile. Or maybe its just I'm too removed from the project right now :) 

That said, an OSS/contributor friendly approach might include optimizing-for/not-assuming domain expert status.

Reply:
Added more comments